### ISSUE TITLE
Allow datagrid column headers to wrap

### ISSUE THREAD

**Issue by elycheea** [carbon-design-system/ibm-products#3517]
Originally posted in #1923
**Designer:** @johannasuarez-koval

### Desired behavior
- [ ] When title is too long, wrap the text to two lines and then truncate the rest of the text
  - [ ] Full text should be shown in a tooltip on hover
- [ ] When using **compact** (`xs`??) data tables, the title or content inside the cells will truncate and not wrap onto two lines.

### To investigate
- [ ] Ensure that keyboard users are able to see the full column title even if truncated

![wrappingHeader](https://user-images.githubusercontent.com/5809654/163624168-03443ea7-d707-496b-91ea-a7cc32b0d9e6.gif)

**Comment by elycheea** [comment]
Adding this as a related issue we should look into with datagrid column headers.

Column headers are not reachable with keyboard, there's no way for keyboard users to see the full text of the truncated column headers.

_Originally posted by @thefirstartist in https://github.com/carbon-design-system/ibm-products/issues/3506#issuecomment-1740061765_

**Comment by ljcarot** [comment]
We've been investigating [TanStack ](https://tanstack.com/table/latest)[Table](https://tanstack.com/table/latest), a third-party, open-source offering, which provides extensive data table capabilities surpassing what our Carbon Datagrid offers. It provides much more flexibility and customization. TanStack Table is headless which means it can easily be added alongside Datagrid component in your product or application. The benefits of more flexibility for product teams and less maintenance for Carbon makes it a win win. Lastly, it is available in multiple frameworks including [React](https://github.com/carbon-design-system/tanstack-carbon/tree/main/react#readme) and [Web Component](https://github.com/carbon-design-system/tanstack-carbon/tree/main/web-components#readme) so it provides an option to non-React product teams.

For these reasons, we have decided to transition from building our own custom table component to using an example-based approach with TanStack Table. Datagrid will still exist in our library for existing teams but we are announcing the deprecation* of the Datagrid component in v2.54.0 release so teams can begin to work through the transition. Details about how to use both Datagrid and TanStack together can be found [here](https://github.com/carbon-design-system/tanstack-carbon/tree/main). 

*Deprecation means that no new features will be added however sev 1 and sev 2 bugs will be supported.
